File: go/stubpromotion.go
// ABOUTME: StubPromotion is a test stub for the Promotion interface.
// ABOUTME: Allows configurable behavior for testing the promotion engine.
package promotionengine

// StubPromotion is a test stub implementation of the Promotion interface.
// It allows you to configure its behavior for testing purposes.
type StubPromotion struct {
	PromoID         PromotionID
	DiscountAmount  Money
	ShouldApply     bool
	Target          string // "line" or "cart"
	DiscountDetails string
}

// ID returns the promotion identifier.
func (s *StubPromotion) ID() PromotionID {
	return s.PromoID
}

// IsApplicable returns whether this promotion should apply.
func (s *StubPromotion) IsApplicable(cart Cart, context PricingContext) bool {
	return s.ShouldApply
}

// Apply returns a configured discount if applicable.
func (s *StubPromotion) Apply(cart Cart, context PricingContext) []AppliedDiscount {
	if !s.ShouldApply {
		return []AppliedDiscount{}
	}

	discount := NewAppliedDiscount(s.PromoID, s.DiscountAmount, s.Target, s.DiscountDetails)
	return []AppliedDiscount{discount}
}

File: java/src/test/java/com/promoengine/promotions/StubPromotion.java
// ABOUTME: Test helper that implements Promotion with configurable behavior.
// ABOUTME: Allows testing of promotion engine without implementing real promotion logic.
package com.promoengine.promotions;

import com.promoengine.domain.Cart;
import com.promoengine.domain.PromotionId;
import com.promoengine.pricing.AppliedDiscount;
import com.promoengine.pricing.PricingContext;

import java.util.List;

public class StubPromotion implements Promotion {

    private final PromotionId promotionId;
    private final boolean shouldApply;
    private final List<AppliedDiscount> discounts;

    public StubPromotion(PromotionId promotionId, boolean shouldApply, List<AppliedDiscount> discounts) {
        this.promotionId = promotionId;
        this.shouldApply = shouldApply;
        this.discounts = discounts != null ? List.copyOf(discounts) : List.of();
    }

    @Override
    public PromotionId id() {
        return promotionId;
    }

    @Override
    public boolean isApplicable(Cart cart, PricingContext context) {
        return shouldApply;
    }

    @Override
    public List<AppliedDiscount> apply(Cart cart, PricingContext context) {
        return discounts;
    }
}
